### PR TITLE
[BEAM-850] Adding display data to sink/sources PubSub, BigQuery, NativeFile, Fil…

### PR DESCRIPTION
**Pull request by pabloem**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:
- [x] Make sure the PR title is formatted like:
  `[BEAM-<Jira issue #>] Description of pull request`
- [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
     Travis-CI on your fork and ensure the whole test matrix passes).
- [x] Replace `<Jira issue #>` in the title with the actual Jira issue
     number, if there is one.
- [ ] If this contribution is large, please file an Apache
     [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).
---

…eBasedSource; Write, Read, ParDo transforms; DoFn.

### CODE REVIEW

File: sdks/python/apache_beam/io/bigquery.py [hunk 1]
```diff
@@ -335,6 +336,24 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
     self.validate = validate
     self.coder = coder or RowAsDictJsonCoder()
 
+  def display_data(self):
+    res = {}
+    if self.query is not None:
+      res['query'] = DisplayDataItem(self.query, label='Query')
```
Comment (robertwb) [inline sdks/python/apache_beam/io/bigquery.py]: What's the difference between label and key? Will users almost always have both? Do we need both?

Comment (pabloem) [inline sdks/python/apache_beam/io/bigquery.py]: *if users do not specify a label, the UI will show the key of the item.

**Comment by pabloem** [comment]
r: @swegner since you added display data to a bunch of components, can you review my PR? : )

File: sdks/python/apache_beam/io/bigquery.py (hunk 1 above)
Comment (swegner) [inline sdks/python/apache_beam/io/bigquery.py]: Phrased differently: 

* The namespace+key make up the URN for a display data item, useful for UI's to provide richer UI for well-known items (for example: recognize a GCS input path and provide a deep-link)
* The label is the human-readable name of the display item used in UI's. If not explicitly provided, defaults to the key.

File: sdks/python/apache_beam/transforms/display.py [hunk 2]
```diff
@@ -86,13 +86,17 @@ def _populate_items(self, display_data_dict):
     """ Populates the list of display data items.
     """
     for key, element in display_data_dict.items():
+      if element is None:
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/display.py]: In the Java SDK, we take a strict stance against registering null display data. Implicitly dropping null values in the SDK can lead to confusing behavior (_"I registered this display item, why isn't it showing up?"_).  Instead, the Java SDK has a helper method `.addIfNotNull(..)`, and the default `.add(..)` method throws if a null value is provided. This makes it very obvious whether you can reliably expect your display item to make it to the UI and when it might be dropped.

I recommend following this semantic here in the spirit of "element of least-surprise", but also so we are consist across SDKs.

File: sdks/python/apache_beam/transforms/display_test.py [hunk 3]
```diff
@@ -22,8 +22,16 @@
...
 import unittest
 
+import hamcrest as hc
+
 import apache_beam as beam
-from apache_beam.transforms.display import HasDisplayData, DisplayData, DisplayDataItem
+from apache_beam.transforms.display import *
+
+
+def make_nspace_display_data(component):
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/display_test.py]: It looks like only a couple tests classes are using this-- should they all?

Note that the transforms and tests we have in the core SDK are likely to get copy-pasted many times over their life, for future SDK transforms and by users. Your implementation here establishes the recipe for how to implement and test display data.

File: sdks/python/apache_beam/transforms/ptransform_test.py [hunk 4]
```diff
@@ -663,6 +665,75 @@ def process(self, context):
...
+    ########################
+    tr = beam.Map(len)
+    dd = DisplayData.create_from(tr)
+    nspace = 'apache_beam.transforms.core.CallableWrapperDoFn'
+    expected_item = DisplayDataItem('len', key='fn',
+                                    label='Transform Function',
+                                    namespace=nspace)
+    hc.assert_that(dd.items, hc.has_item(expected_item))
+
+    ########################
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/ptransform_test.py]: Each of these headings seems to be a separate test case.

File: sdks/python/setup.py [hunk 5]
```diff
@@ -93,6 +93,7 @@ def get_version():
     'protorpc>=0.9.1,<0.12',
     'python-gflags>=2.0,<4.0.0',
     'pyyaml>=3.10,<4.0.0',
+    'pyhamcrest>=1.9,<2.0',
```
Comment (swegner) [inline sdks/python/setup.py]: @robertwb is this the list of dependencies that gets installed when submitting a Dataflow job? Do we want pyhamcrest here?

File: sdks/python/apache_beam/transforms/core.py [hunk 6]
```diff
@@ -29,7 +29,7 @@
 from apache_beam.internal import util
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/core.py]: In Java the list of PTransforms with display data was much longer. Some components I didn't see here:

* Combine
* Window
* AvroIO
* XmlSource / XmlSink
* PipelineOptions

File: sdks/python/apache_beam/transforms/display_test.py (hunk 3 above)
Comment (swegner) [inline sdks/python/apache_beam/transforms/display_test.py]: And on that note, I'll point out how having `DisplayDataMatchers` have been valuable in the Java SDK to lower the bar to writing concise display data tests. Some things that are difficult without them:

* Cannot easily assert on single part of display data. For example, if you have many tests of a PTransform that vary a single property to validate display data, each assertion will redundantly include the full set of display data items + properties. For example, see [`TextIOTest`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java#L538) `testWriteDisplayData()`, `testWriteDisplayDataValidateThenHeader()`, `testWriteDisplayDataValidateThenFooter()`.
* Tests for sub-component display data break encapsulation. For example, in order to test that the `Read` transform properly includes display data for its `Source`, your assert must list all display data for the `Source`. Later updating the `Source` display data will also break the `Read` transform test. In Java we have a `includesDisplayDataFrom(..)` matcher. See [`ReadTest.testDisplayData()`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java#L99)
* Equality-tests on the display data map means adding additional elements to the display data schema (like `path`) will break all tests. It's convenient to assert on only the fields that a PTransform actually sets and exclude others filled in by the SDK (namespace and path).

File: sdks/python/apache_beam/transforms/ptransform_test.py [hunk 7]
```diff
@@ -663,6 +665,75 @@ def process(self, context):
     self.check_label(beam.ParDo(MyDoFn()), r'ParDo(MyDoFn)')
 
 
+class PTransformTestDisplayData(unittest.TestCase):
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/ptransform_test.py]: In Java, we also wrote some "integration" tests to ensure that display data gets properly attached to a pipeline job. There are many ways this can fail that aren't captured in unit tests: 

* Transforms will often wrap or adapt their components, and must ensure that the display data is also properly wrapped.
* Runners can replace transforms in a pipeline graph and must provide equivalent display data in the replacement.
* Composite transforms are lots during pipeline translation and thus composite display data must be passed down into leaf transforms.

In Java SDK, we selected a small subset of "important" transforms to write display data integration tests for-- mostly IO connectors. See [`ReadTest.testBoundedPrimitiveDisplayData()`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java#L110)

File: sdks/python/apache_beam/io/bigquery.py [hunk 8]
```diff
@@ -335,6 +336,24 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
     self.validate = validate
     self.coder = coder or RowAsDictJsonCoder()
 
+  def display_data(self):
```
Comment (pabloem) [inline sdks/python/apache_beam/io/bigquery.py]: This is a use case that I considered too. We could rely on the `mro()` method to obtain display data from super-classes - though we might not always want this (?). What do you think, @robertwb ?

File: sdks/python/apache_beam/io/filebasedsource_test.py [hunk 9]
```diff
@@ -222,6 +225,15 @@ def test_fully_read_single_file(self):
     range_tracker = fbs.get_range_tracker(None, None)
     read_data = [record for record in fbs.read(range_tracker)]
     self.assertItemsEqual(expected_data, read_data)
+    dd = DisplayData.create_from(fbs)
+    nspace = '{}.{}'.format(fbs.__module__, fbs.__class__.__name__)
+    expected_items = [DisplayDataItem(file_name, key='filePattern',
+                                      label="File Pattern", namespace=nspace),
+                      DisplayDataItem('_CompressionType(auto)',
```
Comment (pabloem) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: Done

File: sdks/python/apache_beam/io/bigquery.py [hunk 10]
```diff
@@ -335,6 +336,24 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
...
+
+    if self.table_reference is not None:
+      tableSpec = '{}.{}'.format(self.table_reference.datasetId,
+                                 self.table_reference.tableId)
+      if self.table_reference.projectId is not None:
+        tableSpec = '{}:{}'.format(self.table_reference.projectId,
+                                   tableSpec)
+      res['table'] = DisplayDataItem(tableSpec, label='Table')
+
+    if not self.validate:
```
Comment (pabloem) [inline sdks/python/apache_beam/io/bigquery.py]: Done

File: sdks/python/apache_beam/io/filebasedsource_test.py [hunk 11]
```diff
@@ -501,6 +513,16 @@ def test_source_creation_fails_for_non_number_offsets(self):
       SingleFileSource(
           fbs, file_name='dummy_file', start_offset=None, stop_offset=100)
 
+    dd = DisplayData.create_from(fbs)
```
Comment (pabloem) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: Done

File: sdks/python/apache_beam/io/iobase.py [hunk 12]
```diff
@@ -708,6 +714,10 @@ def __init__(self, *args, **kwargs):
     super(Write, self).__init__(label)
     self.sink = sink
 
+  def display_data(self):
+    return {'sink': self.sink.__class__,
+            'sink_dd': self.sink}
```
Comment (pabloem) [inline sdks/python/apache_beam/io/iobase.py]: Done.

File: sdks/python/apache_beam/transforms/display.py (hunk 2 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/display.py]: That makes sense. I'll consult with Robert.

File: sdks/python/apache_beam/transforms/ptransform_test.py (hunk 4 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/ptransform_test.py]: Done.

File: sdks/python/apache_beam/transforms/core.py (hunk 6 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/core.py]: Yup. I'll work on those in a following transaction.

File: sdks/python/apache_beam/io/pubsub.py [hunk 13]
```diff
@@ -54,6 +55,13 @@ def format(self):
     """Source format name required for remote execution."""
     return 'pubsub'
 
+  def display_data(self):
+    return {'idLabel': DisplayDataItem(self.id_label,
+                                       label='ID Label Attribute'),
+            'topic': DisplayDataItem(self.topic, label='Pubsub Topic'),
+            'subscription': DisplayDataItem(self.subscription,
```
Comment (pabloem) [inline sdks/python/apache_beam/io/pubsub.py]: Done.

File: sdks/python/apache_beam/io/fileio_test.py [hunk 14]
```diff
@@ -631,7 +660,16 @@ def test_write_text_bzip2_file(self):
     with bz2.BZ2File(self.path, 'r') as f:
       self.assertEqual(f.read().splitlines(), self.lines)
 
-  def test_write_text_bzip2_file_auto(self):
+    nspace, dd = make_nspace_display_data(sink)
+    expected_items = [
+        DisplayDataItem('{}{}'.format(self.path, '-SSSSS-of-NNNNN'),
```
Comment (pabloem) [inline sdks/python/apache_beam/io/fileio_test.py]: I looked at the code for the sources and couldn't find where the file name is created. It seems that it's just sent to the dataflow service. I think this should be tested in an integration test. - But I've added an extra test case to keep track of this in case it changes.

**Comment by pabloem** [comment]
@swegner I have addressed your comments. Here's what I added and left for later:

* I **added** an `ItemMatcher` that helps simplify unit tests for display data. I have also added the `ItemMatcher` to our current tests.
* I **added** functions `DisplayDataItem.drop_if_none()`, and `DisplayDataItem.drop_if_default(def_val)` to be consistent with the model in the Java SDK.
* I **added** an extra text case for file pattern prefix/suffix treatment in `NativeFileSink`. We have only two cases at the moment.
* I did **not add** exhaustive test cases for file pattern prefix/suffix treatment in `NativeFileSink`. The code for that seems to not be part of the Python SDK.
* I did **not add** small "integration tests" to make sure display data gets added - I'd like to leave this for a later change (after I've thoroughly studied Java's [`ReadTest.testBoundedPrimitiveDisplayData()`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java#L110)).

Please review, and let me know what you think.
-P. : )

File: sdks/python/apache_beam/io/bigquery.py [hunk 15]
```diff
@@ -335,6 +336,23 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
     self.validate = validate
     self.coder = coder or RowAsDictJsonCoder()
 
+  def display_data(self):
+    res = {}
+    if self.query is not None:
+      res['query'] = DisplayDataItem(self.query, label='Query')
```
Comment (swegner) [inline sdks/python/apache_beam/io/bigquery.py]: Convert to use `.drop_if_none()`, to establish the convention.

File: sdks/python/apache_beam/transforms/display_test.py [hunk 16]
```diff
@@ -22,12 +22,76 @@
...
+import hamcrest as hc
+from hamcrest.core.base_matcher import BaseMatcher
+
 import apache_beam as beam
 from apache_beam.transforms.display import HasDisplayData
 from apache_beam.transforms.display import DisplayData
 from apache_beam.transforms.display import DisplayDataItem
 
 
+def make_nspace_display_data(component):
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/display_test.py]: pydoc

File: sdks/python/apache_beam/transforms/display_test.py [hunk 17]
```diff
@@ -22,12 +22,76 @@
...
+  dd = DisplayData.create_from(component)
+  nspace = '{}.{}'.format(component.__module__, component.__class__.__name__)
+  return nspace, dd
+
+
+class ItemMatcher(BaseMatcher):
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/display_test.py]: pydoc class, and public methods

File: sdks/python/apache_beam/transforms/display_test.py [hunk 18]
```diff
@@ -22,12 +22,76 @@
...
 import apache_beam as beam
 from apache_beam.transforms.display import HasDisplayData
 from apache_beam.transforms.display import DisplayData
 from apache_beam.transforms.display import DisplayDataItem
 
 
+def make_nspace_display_data(component):
+  dd = DisplayData.create_from(component)
+  nspace = '{}.{}'.format(component.__module__, component.__class__.__name__)
+  return nspace, dd
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/display_test.py]: The usage of this method for assertions feels strange because half of the output (`nspace`) feeds into the 'expected' display data, and the other half (`dd`) represents the 'actual'. It doesn't seem intuitive-- without reading the implementation I might expect the return `nspace` to be pulled from the generated display data and represent the 'actual nspace'.

Is it necessary to validate `nspace` on each transform display data test? Since this is generated by the SDK rather than from inputs I think we should only need one test in the display data unit tests and then disregard it for other validations.

File: sdks/python/apache_beam/io/bigquery_test.py [hunk 19]
```diff
@@ -111,6 +114,39 @@ def test_invalid_json_neg_inf(self):
 
 class TestBigQuerySource(unittest.TestCase):
 
+  def test_display_data_item_on_validate_true(self):
+    source = beam.io.BigQuerySource('dataset.table', validate=True)
+
+    nspace, dd = make_nspace_display_data(source)
+    expected_items = [
+        ItemMatcher.matches_kvn('validation', True, nspace),
```
Comment (swegner) [inline sdks/python/apache_beam/io/bigquery_test.py]: Display data tests tend to live among other tests within a component test class. As such, the imported `ItemMatcher` class isn't obviously display-data-specific in context. Is there a better name that makes it contextually obvious that `ItemMatcher` is specific to display data?  Perhaps `DisplayDataMatcher.item_matches(..)`?

File: sdks/python/apache_beam/transforms/core.py [hunk 20]
```diff
@@ -580,6 +590,12 @@ def make_fn(self, fn):
   def process_argspec_fn(self):
     return self.fn.process_argspec_fn()
 
+  def display_data(self):
+    # TODO: Add special handling of callables
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/core.py]: Are these TODO's still relevant?

File: sdks/python/apache_beam/transforms/display_test.py [hunk 21]
```diff
@@ -22,12 +22,76 @@
...
+  def describe_to(self, description):
+    if 'key' in self._match_props:
+      description.append('key is {} '.format(self.key))
+    if 'value' in self._match_props:
+      description.append('value is {} '.format(self.value))
+    if 'namespace' in self._match_props:
+      description.append('namespace is {} '.format(self.namespace))
+
+  @classmethod
+  def matches_key(cls, key):
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/display_test.py]: So far I only see usage of `matches_kvn`. I recommend removing the additional APIs you aren't using-- it's easier to start with a small API and grow it later than to start with a wide API and narrow it.

File: sdks/python/apache_beam/transforms/display.py [hunk 22]
```diff
@@ -145,6 +148,40 @@ def __init__(self, value, url=None, label=None,
     self.value = value
     self.url = url
     self.label = label
+    self._drop_if_none = False
+    self._drop_if_default = False
+
+  def drop_if_none(self):
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/display.py]: Add unit tests for `drop_if_none()` and `drop_if_default()`

File: sdks/python/apache_beam/transforms/ptransform_test.py [hunk 23]
```diff
@@ -663,6 +665,80 @@ def process(self, context):
...
+  def test_map_named_function(self):
+    ########################
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/ptransform_test.py]: Now that these are separate tests, do these block comments add any value? If not, remove.

File: sdks/python/tox.ini [hunk 24]
```diff
@@ -27,6 +27,7 @@ select = E3
 deps=
   pep8
   pylint
+  pyhamcrest
```
Comment (swegner) [inline sdks/python/tox.ini]: Now that this `pyhamcrest` is added here, should it be removed from `setup.py`?

File: sdks/python/apache_beam/transforms/ptransform_test.py (hunk 23 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/ptransform_test.py]: Done.

File: sdks/python/apache_beam/transforms/display.py (hunk 22 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/display.py]: Done.

File: sdks/python/apache_beam/transforms/display_test.py (hunk 17 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/display_test.py]: Done. (`_matches`, and `describe_to` methods come from the super class.)

File: sdks/python/apache_beam/io/bigquery.py (hunk 15 above)
Comment (pabloem) [inline sdks/python/apache_beam/io/bigquery.py]: Done.

File: sdks/python/apache_beam/transforms/display_test.py (hunk 18 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/display_test.py]: Makes sense. I'm on it.

File: sdks/python/apache_beam/transforms/core.py (hunk 20 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/core.py]: I removed that TODO.

File: sdks/python/tox.ini (hunk 24 above)
Comment (swegner) [inline sdks/python/tox.ini]: My understanding of `setup.py` is that it defines how the Beam SDK gets installed, and thus also  dictates the list of dependencies installed on worker VMs to execute Beam pipelines. If that's the case, we should probably keep the dependency list as trim as possible.

@robertwb / @chamikaramj, can you shed light on whether this is the proper way to add a test dependency on `pyhamcrest`?

Comment (chamikaramj) [inline sdks/python/tox.ini]: Probably we should keep this dependency in setup.py as well to support 'python setup.py test' command. @charlesccychen to confirm.

Comment (charlesccychen) [inline sdks/python/tox.ini]: I suggest keeping it in both places and adding this to test_requires in setup.py.  Unfortunately, this doesn't interact well with pip (https://github.com/pypa/pip/issues/1197), but semantically, that is the correct place to put it and it will make python setup.py test work.

**Comment by pabloem** [comment]
I have addressed all comments, and unit tests are passing; but I am having jobs fail on the Dataflow service with these changes. : (
I'm looking into it.

File: sdks/python/setup.py [hunk 25]
```diff
@@ -93,6 +93,10 @@ def get_version():
...
+    ]
+REQUIRED_TEST_PACKAGES = [
+    'pyhamcrest>=1.9,<2.0',
```
Comment (charlesccychen) [inline sdks/python/setup.py]: I suggest that we keep this pyhamcrest and remove the one in REQUIRED_PACKAGES above.

Comment (pabloem) [inline sdks/python/setup.py]: Done.

File: sdks/python/apache_beam/io/filebasedsource_test.py [hunk 26]
```diff
@@ -482,8 +496,8 @@ def setUp(self):
   def test_source_creation_fails_for_non_number_offsets(self):
     start_not_a_number_error = 'start_offset must be a number*'
     stop_not_a_number_error = 'stop_offset must be a number*'
-
-    fbs = LineSource('dymmy_pattern')
+    file_name = 'dymmy_pattern'
```
Comment (charlesccychen) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: driveby nit: "dymmy_pattern" is misspelled throughout.

File: sdks/python/apache_beam/transforms/display_test.py [hunk 27]
```diff
@@ -137,14 +124,39 @@ def display_data(self):
...
+                'non_val': DisplayDataItem(None).drop_if_none(),
+                'def_val': DisplayDataItem(True).drop_if_default(True),
+                'nodef_val': DisplayDataItem(True).drop_if_default(False)}
+
+    dd = DisplayData.create_from(MyDoFn())
+    expected_items = [DisplayDataItemMatcher.matches_kv('some_val',
+                                                        'something'),
+                      DisplayDataItemMatcher.matches_kv('nodef_val',
+                                                        True)]
+    hc.assert_that(dd.items, hc.contains_inanyorder(*expected_items))
```
Comment (swegner) [inline sdks/python/apache_beam/transforms/display_test.py]: Does this validate that the `dd.items` contains *only* the specified items?  You might need to add an additional assertion `hc.assert_that(dd.items, hc.not(hc.contains_inanyorder(..`

Comment (pabloem) [inline sdks/python/apache_beam/transforms/display_test.py]: `contains_inanyorder` [should match the entire sequence](https://pypi.python.org/pypi/PyHamcrest). But just in case, I added the check. : )

**Comment by pabloem** [comment]
Merging with HEAD of `python-sdk` branch fixed issue running jobs on Dataflow. @robertwb can you take a look and merge if appropriate? : )

File: sdks/python/apache_beam/io/bigquery.py (hunk 8 above)
Comment (robertwb) [inline sdks/python/apache_beam/io/bigquery.py]: I think it'd be good practice to call super (consistently).

File: sdks/python/apache_beam/io/bigquery.py [hunk 28]
```diff
@@ -335,6 +336,22 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
...
     self.coder = coder or RowAsDictJsonCoder()
 
+  def display_data(self):
+    res = {'query':
+           DisplayDataItem(self.query, label='Query').drop_if_none(),
+           'validation': DisplayDataItem(self.validate,
+                                         label='Validation Enabled')}
+
+    if self.table_reference is not None:
+      tableSpec = '{}.{}'.format(self.table_reference.datasetId,
```
Comment (robertwb) [inline sdks/python/apache_beam/io/bigquery.py]: Put this in an else clause of the if below.

File: sdks/python/apache_beam/io/bigquery.py [hunk 29]
```diff
@@ -335,6 +336,22 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
     self.validate = validate
     self.coder = coder or RowAsDictJsonCoder()
 
+  def display_data(self):
+    res = {'query':
```
Comment (robertwb) [inline sdks/python/apache_beam/io/bigquery.py]: Is query mutually exclusive with table_reference?

File: sdks/python/apache_beam/io/bigquery_test.py [hunk 30]
```diff
@@ -134,6 +178,15 @@ def test_specify_query_without_table(self):
 
 class TestBigQuerySink(unittest.TestCase):
 
+  def test_schema_descriptor_display_data(self):
+    sink = beam.io.BigQuerySink(
+        'dataset.table', schema='s:STRING, n:INTEGER')
```
Comment (robertwb) [inline sdks/python/apache_beam/io/bigquery_test.py]: Doesn't seem to test the schema.

File: sdks/python/apache_beam/io/bigquery_test.py [hunk 31]
```diff
@@ -143,9 +196,18 @@ def test_parse_schema_descriptor(self):
         field.name: field.type for field in sink.table_schema.fields}
     self.assertEqual({'n': 'INTEGER', 's': 'STRING'}, result_schema)
 
+  def test_project_table_schema_display_data(self):
```
Comment (robertwb) [inline sdks/python/apache_beam/io/bigquery_test.py]: same

File: sdks/python/apache_beam/io/fileio_test.py [hunk 32]
```diff
@@ -592,6 +602,41 @@ def test_write_text_file(self):
...
+        DisplayDataItemMatcher.matches_kv(
+            'filePattern',
+            '{}{}'.format(self.path, '-SSSSS-of-NNNNN')),
+        DisplayDataItemMatcher.matches_kv(
+            'compression',
+            'auto')]
+    hc.assert_that(dd.items, hc.contains_inanyorder(*expected_items))
+
+  def test_text_file_display_data_added_suffix(self):
+    sink = fileio.NativeTextFileSink('{}.{}'.format(self.path, 'pdf'))
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio_test.py]: This test seems to be inconsistent whether the period belongs to the extension or the format pattern.

File: sdks/python/apache_beam/transforms/display_test.py [hunk 33]
```diff
@@ -72,50 +123,56 @@ def display_data(self):
...
-
-    self.assertEqual(set(dd.items), expected_items)
-
-  def test_subcomponent(self):
-    class SpecialParDo(beam.PTransform):
-      def __init__(self, fn):
-        self.fn = fn
-
+    expected_items = [
+        DisplayDataItemMatcher.matches_kvn('complex_url',
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/display_test.py]: Aren't these tests now weaker, as they don't test the short values or labels?

File: sdks/python/apache_beam/transforms/display_test.py (hunk 27 above)
Comment (robertwb) [inline sdks/python/apache_beam/transforms/display_test.py]: The check is redundant, better to omit it.

File: sdks/python/apache_beam/io/bigquery.py (hunk 29 above)
Comment (pabloem) [inline sdks/python/apache_beam/io/bigquery.py]: Yes . If both are set or both are unset, the constructor raises `ValueError`.

File: sdks/python/apache_beam/io/bigquery.py (hunk 28 above)
Comment (pabloem) [inline sdks/python/apache_beam/io/bigquery.py]: Done.

File: sdks/python/apache_beam/io/bigquery_test.py (hunk 30 above)
Comment (pabloem) [inline sdks/python/apache_beam/io/bigquery_test.py]: The schema is not display data in this case. I've removed it in the test.

File: sdks/python/apache_beam/io/bigquery_test.py (hunk 31 above)
Comment (pabloem) [inline sdks/python/apache_beam/io/bigquery_test.py]: Same ; )

File: sdks/python/apache_beam/io/fileio_test.py (hunk 32 above)
Comment (pabloem) [inline sdks/python/apache_beam/io/fileio_test.py]: In this case, the period is part of the pattern. I added it for the future, in case we decide to be smart about parsing the file extensions. The next test (`test_text_file_display_data_suffix`) actually provides a pattern+suffix.

File: sdks/python/apache_beam/transforms/display_test.py (hunk 33 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/display_test.py]: You're right. I've added the matchers for this.

File: sdks/python/apache_beam/transforms/display_test.py (hunk 27 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/display_test.py]: Okay. Done.

File: sdks/python/apache_beam/transforms/display_test.py [hunk 34]
```diff
@@ -78,6 +89,16 @@ def matches_kvn(cls, key, value, namespace):
                namespace=namespace,
                match_props=['key', 'value', 'namespace'])
 
+  @classmethod
+  def matches_all(cls, key, value, namespace, label, shortValue):
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/display_test.py]: Using match_props and adding a new method for every subset doesn't seem very extensible. What if instead you just had

```
class DisplayDataItemMatcher(...):
  IGNORED = object()
  def __init__(self, key, value=IGNORED, namespace=IGNORED, ...)
    ...

```

and so DisplayDataItemMatcher(...) simply matched the keywords passed.

File: sdks/python/tox.ini (hunk 24 above)
Comment (robertwb) [inline sdks/python/tox.ini]: Agreed.

File: sdks/python/apache_beam/io/bigquery.py (hunk 29 above)
Comment (robertwb) [inline sdks/python/apache_beam/io/bigquery.py]: In that case, I'd put it in the else clause to make things clearer.

File: sdks/python/apache_beam/io/fileio_test.py (hunk 32 above)
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio_test.py]: OK, what is this testing in that case? I'd leave the pdf off if it's not really part of the test, as it confuses things.

File: sdks/python/apache_beam/io/fileio_test.py [hunk 35]
```diff
@@ -631,7 +689,20 @@ def test_write_text_bzip2_file(self):
...
+    expected_items = [
+        DisplayDataItemMatcher.matches_kv(
+            'filePattern',
+            '{}{}'.format(self.path, '-SSSSS-of-NNNNN')),
+        DisplayDataItemMatcher.matches_kv(
+            'compression',
+            'bzip2')]
+    hc.assert_that(dd.items, hc.contains_inanyorder(*expected_items))
+
+  def test_write_text_zlib_file_auto(self):
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio_test.py]: zlib?

File: sdks/python/apache_beam/transforms/display_test.py (hunk 34 above)
Comment (pabloem) [inline sdks/python/apache_beam/transforms/display_test.py]: Done.

**Comment by robertwb** [comment]
LGTM, but doesn't seem to rebase cleanly on https://github.com/apache/incubator-beam/commit/ce3f1726ff9c69d718c41cae02cf2d4a370f5a22
